General Clean-up and Fix for #7

JWhitleyAStuff opened the pull request:
This PR generally cleans up package.xml and brings it up to format 2 (defined in http://wiki.ros.org/catkin/package.xml#Format_2_.28Recommended.29); cleans up CMakeLists.txt and applies best-practices produced by the ROS community with regard to ordering and content of add_dependencies, target_link_libraries, and add_compile_options; and renames the message file per ROS naming conventions. Also fixes #7.

JWhitleyAStuff commented:
Now also includes correct CMake dependencies for OpenMP, libunwind, OpenCV, and others. Includes package.xml dependency on libgomp1 - will add libunwind-dev once ros/rosdistro#19615 is approved.

shahvi commented:
When doing a fresh catkin make, I see the error
```
CMake Error at spinnaker_sdk_camera_driver/CMakeLists.txt:42 (find_package):
  By not providing "FindLibUnwind.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "LibUnwind", but CMake did not find one.

  Could not find a package configuration file provided by "LibUnwind" with
  any of the following names:

    LibUnwindConfig.cmake
    libunwind-config.cmake

  Add the installation prefix of "LibUnwind" to CMAKE_PREFIX_PATH or set
  "LibUnwind_DIR" to a directory containing one of the above files.  If
  "LibUnwind" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
See also "/home/vik748/apps/spinnaker_ws/build/CMakeFiles/CMakeOutput.log".
See also "/home/vik748/apps/spinnaker_ws/build/CMakeFiles/CMakeError.log".
Invoking "cmake" failed
```

JWhitleyAStuff commented:
@shahvi - Try pulling, cleaning your catkin workspace, and building again. I added some additional CMake rules just a few minutes ago.

shahvi commented:
@JWhitleyAStuff  still the same error.  I am no expert in CMake, but what we had was working for libunwind.  How does adding all this extra stuff help?
Another point, we did discover that we need to comment out libunwind for an ARM architecture like for a Jetson TX1.  Is this something that can be handled in the CMake rules?

JWhitleyAStuff commented:
@shahvi - The reason these additions are necessary is so that, if libunwind (or one of the other libraries) doesn't exist on the user's system, it is caught at CMake time with a human-readable error instead of by the linker failing with a more cryptic error. This is common practice with ROS as well. The reason for the additions to package.xml is so that rosdep can find the necessary dependencies. This allows clean environments (like CI builds or the ROS build farm) to find the dependencies before compiling. Even though you won't be able to submit this package to the build farm right now (because Spinnaker isn't available as a ROS or Ubuntu package), you can still generate .deb files using CI and use CI to make sure everything builds correctly on each push or PR.

Re: LibUnwind CMake search process, try the latest push. Re: architecture-specific stuff, the FindLibUnwind.cmake seems to take care of that - though, admittedly, I have not tested it in an ARM environment.

shahvi commented:
@JWhitleyAStuff I still have the same error when running catkin_make.

JWhitleyAStuff commented:
@shahvi - Thanks for clarifying that you were using catkin_make. I usually use catkin build and am generally in the spinnaker_sdk_camera_driver folder when I do so. That's why this wasn't showing up for me - I was using CMAKE_SOURCE_DIR which refers to the base of the CMake source infrastructure. In my case, this was the spinnaker folder. In your case, it was the root of the workspace. Using CMAKE_CURRENT_SOURCE_DIR fixed the issue. Please try again.

shahvi commented:
@JWhitleyAStuff , yep I can compile just fine now, thanks for fixing that. I have pushed a commit to the master branch that addresses all the warning that show up with C++11.  Could you please pull forward your PR.
Also, I see that you are depending on OpenMP now, what is that needed for?

JWhitleyAStuff commented:
@shahvi - Strange, I could have sworn that OpenMP was a requirement before. OK. I've removed it and fixed one last C++11 bug in camera.cpp (there is no need to set the CameraPtr object to null - it has a destructor and should take care of cleaning up after itself).

shahvi commented on src/camera.cpp:
```diff
@@ -2,7 +2,6 @@
 
 acquisition::Camera::~Camera() {
 
-    pCam_ = NULL;
```
We need to clear this pointer, otherwise the code spits out an error trace like this on exiting the node:
`7f3381926000-7f3381b25000 ---p 0021f000 08:06 13378414                   /usr/lib/x86_64-linux-gnu/libvpx.so.3.0.0
7f3381b25000-7f3381b27000 r--p 0021e000 08:06 13378414                   /usr/lib/x86_64-linux-gnu/libvpx.so.3.0.0
7f3381b27000-7f3381b28000 rw-p 00220000 08:06 13378414                   /usr/lib/x86_64-linux-gnu/libvpx.so.3.0.0
7f3381b28000-7f3381b2b000 rw-p 00000000 00:00 0   
...
`

JWhitleyAStuff replied:
I reverted this change. This is a very strange (c-style) way to handle a class object but it's their API and we can't really change it so it is what it is.

shahvi commented on src/capture.cpp:
```diff
@@ -289,7 +289,7 @@ void acquisition::Capture::read_parameters() {
     } else ROS_WARN("  'skip' Parameter not set, using default behavior: skip=%d",skip_num_);
 
     if (nh_pvt_.getParam("delay", init_delay_)){
-        if (init_delay_>=0) ROS_INFO("  Init sleep delays set to : %0.2f sec",init_delay_);
```
Init delay can be fraction of seconds, so makes more sense to use float format.

JWhitleyAStuff replied:
Fixed by changing init_delay_ to a float.

JWhitleyAStuff commented:
Very sorry for all of the back-and-forth. I'm going to create a new PR. There are some changes showing up in the PR that shouldn't be there (specifically in camera.cpp). I'll re-apply the changes that actually matter in the new PR.
